system/lib/wasmfs/syscalls.cpp

```diff
@@ -861,6 +861,58 @@ int __syscall_rmdir(intptr_t path) {
   return __syscall_unlinkat(AT_FDCWD, path, AT_REMOVEDIR);
 }
 
+// wasmfs_unmount is similar to __syscall_unlinkat, but assumes AT_REMOVEDIR is true and will unlink nonempty directories.
+int wasmfs_unmount(int dirfd, intptr_t path) {
+  std::string_view p((char*)path);
+  // Ignore trailing '/'.
+  while (!p.empty() && p.back() == '/') {
+    p.remove_suffix(1);
+  }
+  if (p.size() >= 2 && p.substr(p.size() - 2) == std::string_view("/.")) {
+    return -EINVAL;
+  }
+
+  auto parsed = path::parseParent((char*)path, dirfd);
+  if (auto err = parsed.getError()) {
+    return err;
+  }
+  auto& [parent, childNameView] = parsed.getParentChild();
+  std::string childName(childNameView);
+  auto lockedParent = parent->locked();
+  auto file = lockedParent.getChild(childName);
+  if (!file) {
+    return -ENOENT;
+  }
+  // Disallow removing the root directory, even if it is empty.
+  if (file == wasmFS.getRootDirectory()) {
+    return -EBUSY;
+  }
+
+  auto lockedFile = file->locked();
+  if (auto dir = file->dynCast<Directory>()) {
+    if (parent->getBackend() == dir->getBackend()) {
+      // The child is not a valid mountpoint.
+      return -EINVAL;
+    }
+  } else {
+    // A normal file or symlink.
+    return -ENOTDIR;
+  }
+
+  // Cannot unlink/rmdir if the parent dir doesn't have write permissions.
+  if (!(lockedParent.getMode() & WASMFS_PERM_WRITE)) {
+    return -EACCES;
+  }
+
+  // Input is valid, perform the unlink.
+  if (auto err = lockedParent.removeChild(childName)) {
+    return err;
+  }
+
+  // Create a new directory to replace the old mount.
+  return wasmfs_create_directory((char*)path, 0777, parent->getBackend());
```

Comment:
The basic unmount operation should not do this, although perhaps it needs to be done at a higher level for compatibility with the legacy JS API. If so, we should presumably only create this replacement directory if we are overwriting a directory in the first place.

Reply:
I think it might make sense to move the creation of the directory to `js_api.cpp`. Would there be a way to tell if the mountpoint was created by overwriting a directory? Or would it be simpler to directly create a new directory on each call to unmount?

Reply:
Whether a previous directory existed could be determined from the error code returned by the rmdir we do, and specifically we know there was a previous directory only if the rmdir call returns 0.

Reply:
I think the unlink operation in `wasmfs_unmount` might always think a previous directory existed, since the mounted child that we want to unmount will always exist. One idea I had was to add a variable indicating if a previous directory existed to each mounted directory at `mount`, since we can see the result of `rmdir` there, though I'm not sure this is the best approach. Do you have any suggestions for other ways to track whether or not a mount directory "covers" a previous directory?

Reply:
Oh right, I was getting mount and unmount mixed up. Sorry about that. Yes, just unconditionally creating the replacement directory at the JS API level might make sense, if that matches the legacy API behavior.

Perhaps we should talk about whether there is a feasible path toward "fixing" this legacy behavior next time we meet because it really is extremely surprising IMO.